### PR TITLE
Feature: Add multisig client

### PR DESCRIPTION
**Pull request by josemarinas**
## Description

Add multisig client for plugins multisig contract

Task: [APP-1369](https://aragonassociation.atlassian.net/browse/APP-1369)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder of the package after the [UPCOMING] title and before the latest version.
- [ ] I have tested my code on the test network.

[APP-1369]: https://aragonassociation.atlassian.net/browse/APP-1369?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/client/examples/03-encoders/11-add-multisig-members.ts
Comment (brickpop) [inline modules/client/examples/03-encoders/11-add-multisig-members.ts]: Here we are going to need to sync with Michael, because probably a new parameter is expected too. The new minimum amount of signers after adding these new wallets

File: modules/client/examples/04-decoders/15-add-multisig-members.ts
Comment (brickpop) [inline modules/client/examples/04-decoders/15-add-multisig-members.ts]: Same as before, if there's an extra param with the new amount of signers, the struct is going to change

File: modules/client/examples/06-multisig-client/00-installation.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/00-installation.ts]: We also need to pass the minimum amount of approvals needed

File: modules/client/examples/06-multisig-client/04-approve-proposal.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/04-approve-proposal.ts]: We will probably have some parameter to define whether the approver wants to execute the proposal or not. By default, yes

File: modules/client/examples/06-multisig-client/05-can-approve.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/05-can-approve.ts]:
```suggestion
const canVote = await client.methods.canApprove("0x1234567890123456789012345678901234567890");
console.log(canVote);
```
```suggestion
const canApprove = await client.methods.canApprove("0x1234567890123456789012345678901234567890");
console.log(canApprove);
```

File: modules/client/examples/06-multisig-client/05-can-approve.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/05-can-approve.ts]: Don't we need to pass the pluginAddress and the proposal ID? If you already approved a certain proposal you would receive `false`

File: modules/client/examples/06-multisig-client/06-get-members.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/06-get-members.ts]: Do we have a way to know how many approvals are needed? This might look more like `getPluginSettings` than just getting the addresses

File: modules/client/examples/06-multisig-client/04-approve-proposal.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/04-approve-proposal.ts]: We also need another operation: `execute()`. In case someone decides not to pay the execution costs him/herself

File: modules/client/examples/06-multisig-client/07-get-proposal.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/07-get-proposal.ts]:
```suggestion
  approvals: [
```

File: modules/client/examples/06-multisig-client/08-get-proposals.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/08-get-proposals.ts]: Is this actually used, if graphQL doesn't support it?

File: modules/client/src/multisig/interfaces.ts
Comment (brickpop) [inline modules/client/src/multisig/interfaces.ts]: - Add `execute` too

File: modules/client/src/multisig/interfaces.ts
Comment (brickpop) [inline modules/client/src/multisig/interfaces.ts:25]: - Allow to `skipExecution`

File: modules/client/src/multisig/interfaces.ts
Comment (brickpop) [inline modules/client/src/multisig/interfaces.ts]: Also add the new minApprovals value. By default, it is zero (no change)

File: modules/client/src/multisig/interfaces.ts
Comment (brickpop) [inline modules/client/src/multisig/interfaces.ts]: Also add the new minApprovals value. By default, it is zero (no change)

File: modules/client/src/multisig/interfaces.ts
Comment (brickpop) [inline modules/client/src/multisig/interfaces.ts]: plus  `execute()`

File: modules/client/src/multisig/internal/constants.ts
Comment (brickpop) [inline modules/client/src/multisig/internal/constants.ts]: We may also have an `updatePluginSettings` method to set the minimum approval, sync with Michael

File: modules/client/test/helpers/deployContracts.ts
Comment (brickpop) [inline modules/client/test/helpers/deployContracts.ts]: TO DO, right?

File: modules/client/test/helpers/deployContracts.ts
Comment (brickpop) [inline modules/client/test/helpers/deployContracts.ts:183]: What are these 3 values? Multisig only has `minApproval` count now

File: package.json
Comment (brickpop) [inline package.json:23]: Could it make sense to run the plugin tests separately? to avoid loading too much stuff in memory

File: modules/client/examples/03-encoders/11-add-multisig-members.ts
Comment (josemarinas) [reply modules/client/examples/03-encoders/11-add-multisig-members.ts]: Yes, I was plannning on fixing this kind of issues once michael has a first version of the plugin

File: modules/client/examples/06-multisig-client/05-can-approve.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/05-can-approve.ts]: Didn't think of that, I saw it more of a plugin wide thing, but maybe `isMember` would be a better name for what I was thinking of. For this makes sense passing the proposalId. In fact with the new proposalId definition we wont even need the plugin address, just the proposalId

File: modules/client/examples/06-multisig-client/05-can-approve.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/05-can-approve.ts]: :+1:

File: modules/client/examples/06-multisig-client/06-get-members.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/06-get-members.ts]: Yes we can change this with the new parameters, I assumed only addresses will be needed, renaming this to `getPluginSettings` and returning something like this would make more sense:

```
type MultisigPluginSettings {
  addressesOrEns: string[]
  votingSettings: MultisigVotingSettings
}

type MultisigVotingSettings {
  executeOnPass?: boolean // default true
  minSupport: number // 0-1
}
```

Let me know what do you think

File: modules/client/examples/06-multisig-client/08-get-proposals.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/08-get-proposals.ts]: They are going to need it in the app team afaik, for us popularity equals the proposals with more votes, I guess in this case wold be with more approvals?

File: modules/client/src/multisig/internal/constants.ts
Comment (josemarinas) [reply modules/client/src/multisig/internal/constants.ts]: :+1:

File: modules/client/src/multisig/interfaces.ts
Comment (josemarinas) [reply modules/client/src/multisig/interfaces.ts]: :+1:

File: modules/client/src/multisig/interfaces.ts
Comment (josemarinas) [reply modules/client/src/multisig/interfaces.ts:25]: :+1:

File: modules/client/test/helpers/deployContracts.ts
Comment (josemarinas) [reply modules/client/test/helpers/deployContracts.ts]: yep, forgot to add it

File: package.json
Comment (josemarinas) [reply package.json:23]: yes but this would imply changin all the gh workflows

I can add new commands if you want to execute them separately

```json 
"test:erc20": "yarn test client-erc20"
"test:addressList": "yarn test client-addressList"
"test:multisig": "yarn test client-multisig"
```

File: modules/client/examples.md
Comment (brickpop) [inline modules/client/examples.md]: canApprove

File: modules/client/examples.md
Comment (brickpop) [inline modules/client/examples.md:2982]: I personally like `MultisigClient` as opposed to `ClientMultisig` bt this is inconsistent with the rest of client classes. 

If we follow this pattern, then we need to rename the rest to follow the same structure

File: modules/client/examples/03-encoders/11-add-multisig-addresses.ts
Comment (brickpop) [inline modules/client/examples/03-encoders/11-add-multisig-addresses.ts]:
```suggestion
### Add Members (Multisig)
```

File: modules/client/examples/04-decoders/15-add-multisig-addresses.ts
Comment (brickpop) [inline modules/client/examples/04-decoders/15-add-multisig-addresses.ts]: How do we get members + minApprovals when the encoding action only specifies members?

We should be encoding both with the new minApprovals amount

File: modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts
Comment (brickpop) [inline modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts]: Why is this file named `copy`?

File: modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts
Comment (brickpop) [inline modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts]: Why don't we get the addresses to remove as well?

File: modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts
Comment (brickpop) [inline modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts]: I might call it `signers` or `addresses` as opposed to `members`, everywhere

File: modules/client/examples/06-multisig-client/00-installation.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/00-installation.ts]: Does this need to be a bigint? IMO it should be a regular number, and only convert it to a `bigint` internally if needed

File: modules/client/examples/06-multisig-client/07-can-execute.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/07-can-execute.ts]: Does this function require an address? If a proposal is passed, anybody should be able to execute it, only the process ID should be needed here

File: modules/client/examples/06-multisig-client/08-get-plugin-settings.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/08-get-plugin-settings.ts:23]: Isn't it `pluginSettings`? everywhere

File: modules/client/examples/06-multisig-client/10-get-proposals.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/10-get-proposals.ts]: How do we specify what Plugin to list the proposals for? Same applies to everywhere

File: modules/client/src/client-common/interfaces/plugin.ts
Comment (brickpop) [inline modules/client/src/client-common/interfaces/plugin.ts]: As said, is this needed?

**Review by brickpop — COMMENTED** [review]
Next round

File: modules/client/examples/04-decoders/15-add-multisig-addresses.ts
Comment (josemarinas) [reply modules/client/examples/04-decoders/15-add-multisig-addresses.ts]: It follows the naming pattern in the contracts, when you add or remove members you should also specify the new minimum participation. 

Maybe leaving the participation as a percentage in the contracts would make more sense to avoid the need of changing the min approvals each time we add or remove an address

File: modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts
Comment (josemarinas) [reply modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts]: Again, following the contracts naming pattern, if we decide to change it i would do it in both places

File: modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts
Comment (josemarinas) [reply modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts]: This just updates de minimum approvals, why to wue need the addresses to remove?

File: modules/client/examples/06-multisig-client/00-installation.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/00-installation.ts]: Yeah, you are right. I hope there isnt a multisig with more than 9007199254740992 memebers

File: modules/client/examples/06-multisig-client/07-can-execute.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/07-can-execute.ts]: Oh ok, I understood it more like  "can this address execute this proposalId", but I'll change it

File: modules/client/examples/06-multisig-client/08-get-plugin-settings.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/08-get-plugin-settings.ts:23]: What do you mean? Is about the types names `MultisigPluginSettings` is defined but then I use som aliases to make it more readable, I can delete them if you want

File: modules/client/examples/06-multisig-client/10-get-proposals.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/10-get-proposals.ts]: There is a parameter that is `daoAddressOrEns` that we can use to filter by dao. I can add it to the examples or change it by the plugin address however in bot `AddresslistVoting` and `TokenVoting` plugins it uses the dao address to filter and if we decide to change it I would unify it

File: modules/client/examples/06-multisig-client/07-can-execute.ts
Comment (josemarinas) [reply modules/client/examples/06-multisig-client/07-can-execute.ts]: Just reviewed the contracts, it needs the `address` param

File: modules/client/src/client-common/interfaces/plugin.ts
Comment (josemarinas) [reply modules/client/src/client-common/interfaces/plugin.ts]: same, its in the contracts

File: modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts
Comment (brickpop) [reply modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts]: Ping

File: modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts
Comment (brickpop) [reply modules/client/examples/04-decoders/16-remove-multisig-addresses copy.ts]: ok :+1:

File: modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts
Comment (brickpop) [reply modules/client/examples/04-decoders/17-update-multisig-min-approvals.ts]: Sorry, wrong file... :innocent:

File: modules/client/examples/06-multisig-client/07-can-execute.ts
Comment (brickpop) [reply modules/client/examples/06-multisig-client/07-can-execute.ts]: Where do you see the 2 params, @josemarinas ? It only needs the proposalId

https://github.com/aragon/core/pull/167/files#diff-aa6435e11ba402b1ca24319f8bfef661e834e2124a93b4d826586bd8ce2e64a4R286

File: modules/client/examples/06-multisig-client/07-can-execute.ts
Comment (brickpop) [inline modules/client/examples/06-multisig-client/07-can-execute.ts]: Formatting

File: modules/client/examples/06-multisig-client/10-get-proposals.ts
Comment (brickpop) [reply modules/client/examples/06-multisig-client/10-get-proposals.ts]: Whatever the case, it has to be clear how to list the proposals of a certain DAO in the example :)

File: modules/client/src/client-common/interfaces/plugin.ts
Comment (brickpop) [reply modules/client/src/client-common/interfaces/plugin.ts]: Not sure where you see it

https://github.com/aragon/core/pull/167/files#diff-aa6435e11ba402b1ca24319f8bfef661e834e2124a93b4d826586bd8ce2e64a4R286

File: modules/client/src/client-common/utils.ts
Comment (brickpop) [inline modules/client/src/client-common/utils.ts]: eventName?

File: modules/client/examples.md
Comment (brickpop) [inline modules/client/examples.md:2642]: Shouldn't this example go at the beginning, with the rest of DAO creations?

File: modules/client/examples.md
Comment (brickpop) [inline modules/client/examples.md]: (very, very minor) spacing, formatter

File: modules/client/examples.md
Comment (brickpop) [inline modules/client/examples.md]: Warning: the pluginId format is not merged at all. Actually the discussion was closed, then reopened

File: modules/client/examples.md
Comment (josemarinas) [reply modules/client/examples.md:2642]: Mmm we dont have a place to all dao creations.  Each examples folder (tokeVoting, addressListVoting, multisig) have an installation example but each inside its corresponding folder

File: package.json
Comment (brickpop) [reply package.json:23]: We can create a separate task

File: modules/client/test/helpers/deployContracts.ts
Comment (josemarinas) [reply modules/client/test/helpers/deployContracts.ts:183]: thats the version

File: modules/client/examples/03-encoders/13-update-multisig-voting-settings.ts
Comment (brickpop) [inline modules/client/examples/03-encoders/13-update-multisig-voting-settings.ts:19]: Is this already supported?

File: modules/client/test/integration/multisig-client/encoding.test.ts
Comment (brickpop) [inline modules/client/test/integration/multisig-client/encoding.test.ts]:
```suggestion
    it("Should create a Multisig client and an add members action", async () => {
```

File: modules/client/test/integration/multisig-client/encoding.test.ts
Comment (brickpop) [inline modules/client/test/integration/multisig-client/encoding.test.ts]:
```suggestion
    it("Should create a Multisig client and fail to generate a remove members action with an invalid plugin address", async () => {
```

File: modules/client/test/integration/multisig-client/estimation.test.ts
Comment (brickpop) [inline modules/client/test/integration/multisig-client/estimation.test.ts]:
```suggestion
describe("Client Multisig", () => {
```

File: modules/client/test/integration/tokenVoting-client/methods.test.ts
Comment (brickpop) [inline modules/client/test/integration/tokenVoting-client/methods.test.ts:222]: Still the case?

Comment (josemarinas) [reply modules/client/test/integration/tokenVoting-client/methods.test.ts:222]: Yes, I'm not sure if subgraph should have the addresses here but at the moment is empty but the tokens were minted to the correct address

**Review by brickpop — APPROVED** [review]

**Pull request merged**
